(CDAP-1516) Namespace Explore operations

ParthGandhi opened the pull request:
Major design choices: 
1. Split up the Explore handlers into 2 types - those that require a namespace and those that don't. For instance, to submit a query you need a namespace, but to get the status of that query you need only the `QueryHandle` since those are unique across namespaces. 
2. ~~`QueryHandle` now has a `namespaceId` that isn't used for equality checking (since some operations aren't namespaced and you don't have their namespace at that point). However, it is used when getting namespace-specific queries (such as `getQueries`).~~ => Resolved by moving the `namespace` to the value of the cache, which also significantly cleans up the logic.

TODO:
1. ~~PFS needs to be partitioned - sync up with @bdmogal on this. Relevant tests have been ignored.~~ => Fixed.
2. Add more tests for namespaced operations.

anwar6953 commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/FileStreamAdmin.java:
```diff
@@ -452,10 +452,11 @@ private void mutateStates(long groupId, int instances, Set<StreamConsumerState>
     if (cConf.getBoolean(Constants.Explore.EXPLORE_ENABLED)) {
       // It shouldn't happen.
       Preconditions.checkNotNull(exploreFacade, "Explore enabled but no ExploreFacade instance is available");
+      // TODO: Re-enable explore enable/disable when namespacing streams PR is closed.
```
its closed

ParthGandhi replied:
Yeah, I added the code but never removed the TODO..

bdmogal commented on cdap-explore-client/src/main/java/co/cask/cdap/explore/client/ExploreFacade.java:
```diff
@@ -54,73 +55,75 @@ public ExploreFacade(ExploreClient exploreClient, CConfiguration cConf) {
-    handleExploreFuture(futureSuccess, "drop", "partition", name);
+    ListenableFuture<Void> futureSuccess = exploreClient.dropPartition(datasetInstance, key);
+    handleExploreFuture(futureSuccess, "drop", "partition", datasetInstance.getId());
   }
```
Will perhaps have to add methods to create/drop DBs here? maybe later PR.

sreevatsanraman commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java:
```diff
@@ -118,7 +118,7 @@ protected void addPartitionToExplore(PartitionKey key, String path) {
       if (exploreFacade != null) {
         try {
-          exploreFacade.addPartition(getName(), key, files.getLocation(path).toURI().getPath());
+//          exploreFacade.addPartition(getName(), key, files.getLocation(path).toURI().getPath());
```
remove the dead code.

ParthGandhi replied:
Not dead code. This is currently blocked by some changes needed in PFS for explore to work.

ParthGandhi replied:
Yup, will do.

sreevatsanraman commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java:
```diff
@@ -140,7 +140,7 @@ private void dropPartitionFromExplore(PartitionKey key) {
       if (exploreFacade != null) {
         try {
-          exploreFacade.dropPartition(getName(), key);
+//          exploreFacade.dropPartition(getName(), key);
```
Same here.

sreevatsanraman commented on cdap-explore-client/src/main/java/co/cask/cdap/explore/client/ExploreHttpClient.java:
```diff
@@ -329,6 +347,24 @@ public QueryHandle getTypeInfo() throws ExploreException, SQLException {
 
+  @Override
+  public QueryHandle createNamespace(Id.Namespace namespace) throws ExploreException, SQLException {
+    HttpResponse response = doPut(String.format("data/explore/namespaces/%s", namespace.getId()), null, null);
```
namespace here not top level. The API should be 
`namespaces/{namespaceId}/data/explore`

ParthGandhi replied:
As discussed offline, adding a note in `ExploreMetadataHttpHandler` explaining why.

sreevatsanraman commented on cdap-explore-client/src/main/java/co/cask/cdap/explore/client/ExploreHttpClient.java:
```diff
@@ -329,6 +347,24 @@ public QueryHandle getTypeInfo() throws ExploreException, SQLException {
+
+  @Override
+  public QueryHandle deleteNamespace(Id.Namespace namespace) throws ExploreException, SQLException {
+    HttpResponse response = doDelete(String.format("data/explore/namespaces/%s", namespace.getId()));
```
same here.

sreevatsanraman commented on cdap-explore-jdbc/src/main/java/co/cask/cdap/explore/jdbc/ExploreDriver.java:
```diff
@@ -65,17 +65,26 @@ public Connection connect(String url, Properties info) throws SQLException {
 
+    List<String> namespaceParams =
+      Lists.newArrayList(params.getExtraInfos().get(ConnectionParams.Info.NAMESPACE));
+    if (namespaceParams != null && !namespaceParams.isEmpty() && !namespaceParams.get(0).isEmpty()) {
```
namespaceParams cannot be null, since we are just creating it above. Can we remove the check for null?

ParthGandhi commented:
@sreevatsanraman Addressed comments.

In addition, also refactored the `QueryHandle` to not include namespace and move it to the value of the cache instead. This clears up the logic.

ParthGandhi commented:
http://builds.cask.co/browse/CDAP-DUT844-10

bdmogal commented on cdap-client-tests/src/test/java/co/cask/cdap/client/QueryClientTestRun.java:
```diff
@@ -69,7 +71,9 @@ public void testAll() throws Exception {
 
     executeBasicQuery();
-    exploreClient.disableExploreDataset("cdap.default." + FakeApp.DS_NAME).get();
+    Id.DatasetInstance datasetInstance = Id.DatasetInstance.from(getClientConfig().getNamespace(),
```
Any way to set a non-default namespace in the ``ClientConfig`` before this test, then construct the equivalent of ``cdap.default.`` String literal below using a combination of ``ClientConfig`` and ``CConfiguration``?

bdmogal commented on cdap-client/src/main/java/co/cask/cdap/client/QueryClient.java:
```diff
@@ -31,6 +31,7 @@
 public class QueryClient {
 
   private final ExploreClient exploreClient;
+  private final Supplier<String> namespaceSupplier;
```
Why need a supplier?

bdmogal commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceHandler.java:
```diff
@@ -346,15 +341,14 @@ private boolean dropDataset(Id.DatasetInstance datasetInstanceId, DatasetSpecifi
     } catch (Exception e) {
       String msg = String.format("Cannot disable exploration of dataset instance %s: %s",
-                                 name, e.getMessage());
+                                 datasetInstance.getId(), e.getMessage());
```
May be worth it to not to a ``getId`` here, so both namespace and instance name get logged?

bdmogal commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceHandler.java:
```diff
@@ -362,14 +356,14 @@ private void disableExplore(String name) {
     } catch (Exception e) {
       String msg = String.format("Cannot enable exploration of dataset instance %s of type %s: %s",
-                                 name, creationProperties.getProperties(), e.getMessage());
+                                 datasetInstance.getId(), creationProperties.getProperties(), e.getMessage());
```
same comment as above

bdmogal commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/UnderlyingSystemNamespaceAdmin.java:
```diff
@@ -82,5 +92,9 @@ protected void delete(Id.Namespace namespaceId) throws IOException {
     }
+
+    if (cConf.getBoolean(Constants.Explore.EXPLORE_ENABLED)) {
+      exploreFacade.remove(namespaceId);
```
Would it be better to call this method ``removeNamespace``? To keep it consistent with ``createNamespace`` and to perhaps distinguish between removing a table and removing a namespace?

ParthGandhi commented on cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java:
```diff
@@ -863,11 +928,14 @@ public void close(QueryHandle handle) throws ExploreException, HandleNotFoundExc
-          QueryStatus status = getStatus(entry.getKey());
-          result.add(new QueryInfo(entry.getValue().getTimestamp(),
-                                   entry.getValue().getStatement(), entry.getKey(), status, false));
+        // TODO: Check for namespace here too?
```
Need to remove this comment. I do check for namespace here.

bdmogal replied:
Already checking. so comment redundant?

bdmogal commented on cdap-explore/src/main/java/co/cask/cdap/explore/executor/AbstractExploreMetadataHttpHandler.java:
```diff
@@ -0,0 +1,103 @@
+  }
+
+  protected void genericEndpointExecution(HttpRequest request, HttpResponder responder,
+                                        EndpointCoreExecution<Void> execution) {
```
indentation is off

bdmogal commented on cdap-explore/src/main/java/co/cask/cdap/explore/executor/AbstractExploreMetadataHttpHandler.java:
```diff
@@ -0,0 +1,103 @@
+  private static final Gson GSON = new Gson();
+
+  protected void handleResponseEndpointExecution(HttpRequest request, HttpResponder responder,
+                                               final EndpointCoreExecution<QueryHandle> execution) {
```
indentation is off

bdmogal commented on cdap-explore/src/main/java/co/cask/cdap/explore/executor/NamespacedExploreMetadataHttpHandler.java:
```diff
@@ -0,0 +1,146 @@
+    LOG.trace("Received get tables for current user");
+    try {
+      responder.sendJson(HttpResponseStatus.OK, exploreService.getTables(namespaceId));
+    } catch (ExploreException e) {
```
What happens if there are other exceptions? Should probably just catch ``Exception``, if you are sending an ``INTERNAL_SERVER_ERROR``.

bdmogal commented on cdap-explore/src/main/java/co/cask/cdap/explore/executor/NamespacedExploreMetadataHttpHandler.java:
```diff
@@ -0,0 +1,146 @@
+    } catch (TableNotFoundException e) {
+      LOG.error("Could not find table {}", table, e);
+      responder.sendStatus(HttpResponseStatus.NOT_FOUND);
+    } catch (ExploreException e) {
```
same comment as above.

bdmogal commented on cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java:
```diff
@@ -591,7 +595,52 @@ public QueryHandle getTypeInfo() throws ExploreException, SQLException {
+  }
+
+  @Override
+  public QueryHandle execute(String namespace, String statement) throws ExploreException, SQLException {
```
Better to pass ``Id.Namespace`` to keep it consistent

bdmogal commented on cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java:
```diff
@@ -599,8 +648,22 @@ public QueryHandle execute(String statement) throws ExploreException, SQLExcepti
+            public QueryStatus.OpStatus call() throws Exception {
+              return fetchStatus(useDatabaseHandle).getStatus();
+            }
+          }, 5, TimeUnit.SECONDS, 100, TimeUnit.MILLISECONDS);
```
Maybe a jira for making the timeout configurable.

bdmogal commented on cdap-explore/src/test/java/co/cask/cdap/explore/service/HiveExploreServiceStopTestRun.java:
```diff
@@ -39,7 +40,7 @@ public void testServiceStop() throws Exception {
     Set<Long> beforeTxns = transactionManager.getCurrentState().getInProgress().keySet();
 
-    exploreService.execute("show tables");
+    exploreService.execute(Constants.DEFAULT_NAMESPACE, "show tables");
```
Possible to use non-default namespace?

bdmogal commented on cdap-explore/src/test/java/co/cask/cdap/hive/objectinspector/StandardObjectInspectorsTest.java:
```diff
@@ -444,7 +444,6 @@ private void doStandardObjectInspectorTest(boolean testComments) {
 
     // Settable
     Object struct3 = soi1.create();
-    System.out.println(struct3);
```
Nice!

bdmogal commented:
Took a pass at this. Mostly looks good. Minor comments. But we'll probably need one more PR to test the backward compatibility changes introduced by #1408

bdmogal commented:
Also, I did not see any updates to ``DatasetsProxyRule``/``DatasetsProxyRuleTest``. May want to take a look at it.

poornachandra commented on cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java:
```diff
@@ -599,8 +648,22 @@ public QueryHandle execute(String statement) throws ExploreException, SQLExcepti
       SessionHandle sessionHandle = cliService.openSession("", "", sessionConf);
       try {
+        // Switch database to the one being passed in.
+        final OperationHandle useDatabaseHandle = doExecute(sessionHandle, String.format("USE %s", namespace));
```
You should be able to use `SessionState.get().setCurrentDatabase(dbName)` to set the database.

poornachandra replied:
Also, check what happens when an invalid namespace is given. Does the error message look meaningful? If not should we validate the namespace before using it?

ParthGandhi replied:
Will see if that works here.

Yes, the error message is meaningful. However, we do want to do validation of namespaces but that will not happen in this PR. Theres some work to be done around that bit in general for the whole of CDAP.

ParthGandhi commented:
@bdmogal @poornachandra Addressed comments. Please review again.

bdmogal commented:
LGTM.

ParthGandhi commented:
http://builds.cask.co/browse/CDAP-DUT844-13
